README.md
# Condor

[Terraform Module Registry Docs](https://registry.terraform.io/modules/vultr/condor/vultr/latest)

## Table of Contents
* [Requirements](#requirements)
* [Quick Start](#quick-start)
* [Firewall Configuration](#firewall-configuration)
  * [Control Plane HA VLB Firewall](#control-plane-ha-vlb-firewall)
  * [Cluster Nodes Vultr Firewall](#cluster-nodes-vultr-firewall)
* [Extensions](#extensions)
  * [Helm](#helm)
  * [Manifest Bundles](#manifest-bundles)
* [Limitations](#limitations)

## Requirements
  * A funded Vultr account and API Key, should be configured as an environment variable to be consumed by the Terraform provider:
  ```sh
  export VULTR_API_KEY="<api-key-here>"
  ```
  Alternatively define `deployment_vultr_api_key` in your project.
  * The [k0sctl](https://github.com/k0sproject/k0sctl) executable must be installed and in your executable path.

## Cloud Addons
  * Installs the [Vultr CCM](https://github.com/vultr/vultr-cloud-controller-manager)
  * Installs the [Vultr CSI](https://github.com/vultr/vultr-csi)

## Quick Start
Usage and input details can be found in the [Terraform Module Registry Docs](https://registry.terraform.io/modules/vultr/condor/vultr/latest), or use the quickstart below.

1) Create a `main.tf` file with the [Required Inputs](https://registry.terraform.io/modules/vultr/condor/vultr/latest?tab=inputs#required-inputs):
``` hcl
# main.tf

module "condor" {
  source                 = "vultr/condor/vultr"
  version                = "2.0.0"
  cluster_vultr_api_key  = var.cluster_vultr_api_key
  control_plane_firewall_rules = [
    {
      port    = 6443
      ip_type = "v4"
      source  = "0.0.0.0/0"
    }
  ]
}
```
  * The Control Plane Firewall rule in this example exposes the Kubernetes API globally, it is recommended you configure a more restrictive rule or rules on production clusters. 
  * Passing the Cluster API Key as plain text is not recommended for anything beyond testing, use an environment variable as described [here](https://www.terraform.io/docs/cli/config/environment-variables.html#tf_var_name).

2) Configure any [Optional Inputs](https://registry.terraform.io/modules/vultr/condor/vultr/latest?tab=inputs#optional-inputs) if you wish to change from the defaults.

3) Deploy
``` sh
$ terraform init && terraform apply
```

4) The Admin Kubeconfig is written locally to your working directory by default as `admin-<terraform-workspace>.conf`. This may not be desireable in some scenarios, e.g. running Terraform in automation(CI/CD pipelines). If you wish to change this set the `write_kubeconfig` variable to false. If you wish to retreive the kubeconfig manually, you can run the following after your cluster has been deployed:
``` sh
$ k0sctl kubeconfig > /path/to/admin.conf
```

5) Verify cluster functionality
``` sh
kubectl --kubeconfig admin.conf get no 
kubectl --kubeconfig admin.conf get po -n kube-system
```

## Firewall Configuration
### Control Plane HA VLB Firewall
The Control Plane LB Firewall is configured to allow only what is needed by the cluster as described in the [K0s Networking Docs](https://docs.k0sproject.io/v1.22.2+k0s.1/networking/#required-ports-and-protocols) by default. The Kubernetes API will not be accessible without configuring an additional rule or rules(as shown in the quickstart example) via the `control_plane_firewall_rules` input variable. E.g.:
``` hcl
  control_plane_firewall_rules = [
    {
      port    = 6443
      ip_type = "v4"
      source  = "0.0.0.0/0"
    }
  ]
```
As also stated in the quickstart, this example rule exposes the Kubernetes API globally, your rules should be more restrictive for production clusters.

### Cluster Nodes Vultr Firewall
The cluster nodes(control plane and workers) Vultr Firewall defaults to allowing only SSH globally. This is generally acceptable, however if you would like to restrict access further you may disable this rule by setting the `allow_ssh` input variable to `false` then configuring the desired rule/rules outside of this module using the `cluster_firewall_group_id` output in your rules. 

## Extensions
### Helm
Helm Repositories and Charts may be configured/deployed during initial cluster init via the `helm_repositories` and `helm_charts` variables. Note, namespace in a chart definition is required for K0s. Example:
``` hcl
  helm_repositories = [
    {
      name = "argo"
      url  = "https://argoproj.github.io/argo-helm"
    }
  ]
  helm_charts = [
    {
      name      = "argocd"
      chartname = "argo/argo-cd"
      version   = "3.6.8"
      namespace = "argocd"
    }
  ]
```
Please see the [Helm Chart Deployer](https://docs.k0sproject.io/v1.22.2+k0s.1/helm-charts/) docs for a comprehensive list of field/parameter values and further details. Note, this feature entails [Limitations](#limitations).

### Manifest Bundles
You may deploy any Kubernetes manifests automatically with the [K0s Manifest Deployer](hhttps://docs.k0sproject.io/v1.22.2+k0s.1/manifests/#manifest-deployer) by placing your manifests in the `/var/lib/k0s/manifests` directory. Doing so via this module is not supported, however you may use the resulting `controller_ips` module output as arguments to a separate module that copies your manifests to the specified directory(or as stated in the linked K0s docs, a "stack" subdirectory).

## Limitations
* Shrinking of the Control Plane is not supported, only growing. You will need to manually run `k0s etcd leave` on all Control Plane nodes with index > 0 prior to shrinking the `controller_count`. An initial attempt was made to implement this in a destroy time provisioner, however it caused issues when running `terraform destroy` to destroy the entire plan. This may be revisited at a later date. 
* Etcd running on the Control Plane currently goes over the public network due to k0sctl configuration limitaitons - each controllers respective `spec.k0s.spec.storage.peerAddress` parameters would require different values. The Cluster Vultr Firewall allows only Control plane nodes to access the Etcd port(`2380`). Will likely revisit this or look into a PR to K0sctl so that the private network may be used.
* Helm Charts/Repositories are not removed from your cluster if removed from your Terraform configuration. The manifests must be manually removed from each controller in the `/var/lib/k0s/manifests/helm/` directory, and the resources/namespace manually deleted. This makes it less than ideal for continued deployments to your cluster, but great for bootstrapping core applications. A tool such as ArgoCD as used in the Helm example is recommended for long term deployments to your cluster. 


CHANGELOG.md
# Change Log
## v2.x.x - n2k development 
### Breaking Changes
* Using Keys through ssh-agent is dropped, instead a SSH-keypair per deployment is created, additional public SSH-keys can still be added
* Cluster random id is optional through `var.cluster_append_random_id`
* Cluster random id length can be controlled through `var.cluster_random_id_length`
* Bump Terraform provider for Vultr to `2.4.2`
* Vultr API Key for provisioning can be also specified through `var.deployment_vultr_api_key`
### Changes
* Bump K0s release from `v1.21.3+k0s.0` to `v1.22.2+k0s.1`
* Bump kube-system component versions
* Bump vultr-csi and vultr-ccm to `v0.3.0`
* K0s components can be disabled which is supported since `v1.22` through `var.k0s_disable_components`
* Vultr DNS records can be created optionally if `var.cluster_create_external_dns_hosts` and `var.cluster_external_dns_domain` are set
* Vultr CSI can be disabled through `var.enable_vultr_csi`
* Vultr CCM can be disabled through `var.enable_vultr_ccm`
* Switched `apt` to `apt-get` in `scripts/provisioning.sh`
* Added `apt-get dist-upgrade` to `scripts/provisioning.sh`
* Changed sleep-time to 5s in `safe_apt()`
* Added 30s sleep to `scripts/provisioning.sh` on startup, because there are still some update tasks running when vps becomes available
* `$NODE_ROLE` in `scripts/provisioning.sh` is now passed from terraforms `remote-exec` provisioner, instead of regex-ing hostnames
### Features
* HA Control Plane
* Isolated control plane (Control Plane nodes are not part of the cluster)
* Control Plane and Worker Node firewalls.
* Declarative cluster and component (Vultr CCM, CSI, Calico, etc.) upgrades
* Vultr Cloud Controller Manager and Vultr Container Storage Interface are optional
* K0s manifest deployer support
* External DNS records for ha-plane and nodes when specifing a domain

## [v2.0.0](https://github.com/vultr/terraform-vultr-condor/releases/tag/v2.0.0) (2021-09-03)
### Breaking Changes
* Condor is now based on Mirantis K0s rather than Kubeadm, as such v2 is completely incompatible with previous releases.
### Features
* HA Control Plane
* Isolated control plane(Control Plane nodes are not part of the cluster)
* Control Plane and Worker Node firewalls. 
* Declarative cluster and component(Vultr CCM, CSI, Calico, etc.) upgrades
* K0s manifest deployer support 

## [v1.3.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.3.0) (2021-08-25)
### Breaking Changes
* Remove `calico_wireguard` variable - nodes were not properly configured, will need to revisit
* Add `calico_mode` variable - previously defaulted to `vxlan`, is now configurable but defaults to `bird`
### Changes
* Bump K0s release from `v1.21.1+k0s.0` to `v1.21.3+k0s.0`
* Bump kube-system component versions

## [v1.2.3](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.3) (2021-07-14)
### Fixes
* Handle dynamic NIC names.

## [v1.2.2](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.2) (2021-06-21)
### Fixes
* Fix firewall configuration after Vultr image changes. 

## [v1.2.1](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.1) (2021-06-19)
### Changes
* Template Vultr CSI 
* Add Vultr CSI image/version vars
* Add kubeconfig filename tf workspace suffix

## [v1.2.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.0) (2021-06-12)
### Features
* Support K0s Helm deployments.
### Changes
* Convert module internal K0sctl configuration to HCL from YAML.
### Fixes
* Change Controller/Worker network interfaces from ens3/ens7 to enp1s0/enp6s0 due to Vultr image changes. 

## [v1.1.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.1.0) (2021-06-06)
### Features
* Write Kubeconfig locally option.
* Control Plane VLB Firewall. 
### Changes
* Add variable descriptions.
* Lock up cluster firewall, SSH only by default. 
* Docs updates.
### Fixes
* README markdown.

## [v1.0.1](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.0.1) (2021-06-05)
### Fixes
* Remove unused variables from triggers map.

## [v1.0.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.0.0) (2021-06-05)
### First Release


scripts/provision.sh
#!/usr/bin/env bash
set -euxo posix

sleep 30

# disable resolvconf add default vultr nameserver to resolv.conf
# to eliminate coredns loop issues
systemctl disable --now resolvconf.service rdnssd.service
rm /etc/resolv.conf
echo "nameserver 108.61.10.10" > /etc/resolv.conf

safe_apt(){
	while fuser /var/{lib/{dpkg,apt/lists},cache/apt/archives}/lock >/dev/null 2>&1 ; do
		echo "Waiting for apt lock..."
		sleep 5
	done
	apt-get "$@"
}

safe_apt -y update
safe_apt -y upgrade
safe_apt -y dist-upgrade
safe_apt -y install jq

PUBLIC_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[0].mac')
PUBLIC_NIC=$(ip -j link | jq --arg PUBLIC_MAC $PUBLIC_MAC '.[] | select(.address==$PUBLIC_MAC)' | jq -r .ifname)
INTERNAL_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[1].mac')
INTERNAL_NIC=$(ip -j link | jq --arg INTERNAL_MAC $INTERNAL_MAC '.[] | select(.address==$INTERNAL_MAC)' | jq -r .ifname)
INTERNAL_IP=$1
CONTROL_PLANE_PORTS=(6443 2379 2380 10250 10251 10252 8132 8133 9443)

if [ "$NODE_ROLE" = "controller" ] || [ "$NODE_ROLE" = "worker" ]; then
    echo "Configuring host $(hostname) as role: $NODE_ROLE"
    sleep 5
else
    echo "Environment variable NODE_ROLE not set or not equal controller|worker (current value = $NODE_ROLE)"
    echo "aborting..."
    sleep 90
    exit 1
fi

case $NODE_ROLE in
    controller)
        for port in "${CONTROL_PLANE_PORTS[@]}"; do
            ufw allow $port
            ufw allow in on $INTERNAL_NIC
        done
    ;;
    worker)
        ufw allow 10250
        ufw allow 179
        ufw allow 9443
        ufw allow 4789/udp
        ufw allow 8132:8133/tcp
        ufw allow 30000:32767/tcp
        ufw allow in on $INTERNAL_NIC
    ;;
esac

ufw reload

cat <<-EOF > /etc/systemd/network/public.network
  [Match]
  MACAddress=$PUBLIC_MAC

  [Network]
  DHCP=yes
EOF

cat <<-EOF > /etc/systemd/network/private.network
  [Match]
  MACAddress=$INTERNAL_MAC

  [Network]
  Address=$INTERNAL_IP
EOF

echo "# For k0s"                            >> /etc/hosts
echo "$INTERNAL_IP             $(hostname)" >> /etc/hosts

systemctl enable systemd-networkd systemd-resolved
systemctl restart systemd-networkd systemd-resolved
systemctl disable networking
